### ISSUE TITLE
CheckTx

### ISSUE THREAD

**Issue by arrivets** [mosaicnetworks/babble#125]
# CheckTx

Several people have asked us to extend the consensus / app interface with a
CheckTx method to prevent "bad" transactions from being included in the 
hashgraph and blockchain. There are very good reasons for doing so, but the
implementation is trickier than it appears because of the leaderless and 
asychronous nature of the Hashgraph algorithm.   

## What are we trying to achieve with a CheckTx method?

Check that a transaction meets certain requirements (from the perspective of the
app) before further gossiping and including it in the hashgraph. We don't want
garbarge being passed around. As an example, with EVM-Lite, we could verify that
a transaction is a well formed EVM transaction, correctly signed, and that the
sender has enough balance to cover the transaction fee.

## Potential implementation

Before including an Event, in core.Sync, call the CheckTx for each transaction
contained in that event. This is a very easy code change but does it undermine
the consensus algorithm?

## Question 1:

Are different nodes guaranteed to obtain the same results from CheckTx, on the
same event? 

An Event can have multiple children, meaning that multiple nodes may be calling
CheckTx on the same transactions, at different times, and get different results.

**Ex**:

```
e1   |   e2
|  \ | / | 
e0   x   |
|    |   |
|    |   |
A    B   C 
```

Are A and B guaranteed to obtain the same results when checking x's 
transactions? We are obviously interested in the general case, not just this 
specific example.
 
## Question 2

Does question 1 even matter?

Does it matter if nodes compute different results for CheckTx? Intuition tells
us that it does because it opens the door to colluding malicious groups (even 2
would be enough). But the argument remains to be formalised.

**Comment by ethanfrey** [comment]
Thank you for opening this issue. I will comment more later, but in the tendermint model, it doesn't matter if Checktx is the same in each proposer, it may be subjective, while the execution must be objective and deterministic.

In tendermint, a failing Checktx means that it doesn't accept a TX from a client, and it doesn't Re gossip from a peer. It will sign a block that contains failing checktx.

To adopt to babble, we would have to see if that is valid. Certainly rejecting from clients is good. But rejecting gossip may be problematic

**Comment by arrivets** [comment]
Ok, I think there might be a fundamental difference in Babble and Tendermint regarding how txs are submitted to the consensus engine.

In Babble, transactions can only be submitted through the App. So users wishing to send a transaction, have to speak to the App, not Babble, and it's left to the app to filter bad transactions before relaying them. For example EVM-Lite checks the validity of transactions submitted by users, before passing them down to the consensus engine. That's how we reject from clients, and we don't need the CheckTx for this.

```
         +-------+
Tx ----> | APP   | 
         +-------+ 
	   |   |
	 +--------+
         | Babble |
         +--------+
```
Am I correct in thinking that Tendermint accepts transactions directly from the user? 

```
         +-------+
         | APP   | 
         +-------+ 
	   |   |
	 +------------+
Tx ----> | Tendermint |
         +------------+
```

That would explain why Tendemint needs a CheckTx method to reject from clients.

That being said, rejecting from clients doesn't protect against malicious nodes letting garbage through, so there is a case for rejecting gossip as well. Since most of the requests for CheckTx have come from Tendermint users, I assumend this was a feature of Tendermint.

**Comment by ethanfrey** [comment]
> Am I correct in thinking that Tendermint accepts transactions directly from the user?

Yes, the rpc and the mempool are handled in tendermint directly. It sits in front of the App, kind of like Apache. Thus the neccesity there.

> In Babble, transactions can only be submitted through the App. So users wishing to send a transaction, have to speak to the App, not Babble, and it's left to the app to filter bad transactions before relaying them. For example EVM-Lite checks the validity of transactions submitted by users, before passing them down to the consensus engine.

Ahh... I see now that this is well defined in https://github.com/mosaicnetworks/babble#design but I got stuck into assuming there was a similar architecture to Tendermint-ABCI. The actual user-facing API is app-specific then, and Babble just takes care or managing the consensus "under the hood", correct?

I think doing App-based filtering on new Client Tx is a good solution, and seems like it is already done. As to further filtering... if one node is malicious and attempts to gossip garbage, then having a `CheckTx` filter could be a way for other nodes to refuse to make a new Event. I think this could be added without major implications to the consensus layer, and is in parallel to the Tendermint "filter in mempool" approach. However, one a bad Tx is witnessed by a few other members and buried in a tree of Tx,  if a new member refuses to accept the entire tree based on the Tx, this could break the network. And thus this `CheckTx` would only be useful when there was exactly *one* malicious member. Two of them could bury an invalid Tx below various valid Tx and then force it into a block anyway.

I do think some more reflection on protection against malicious member node spamming the network would be useful sometime, I think the main problem is the spam from anonymous clients.

Thank you for the explanation and I am happy to see this Issue resolve with a clear explanation of the need for filtering in the README or other docs. And maybe a "eventual enhancement" tag for protection against spamming members, which definitely requires more work and deeper reflection.

**Comment by arrivets** [comment]
> the actual user-facing API is app-specific then, and Babble just takes care or managing the consensus "under the hood", correct?

Yes, I just committed a change to clarify that in the README and docs.

> ... if one node is malicious and attempts to gossip garbage, then having a CheckTx filter could be a way for other nodes to refuse to make a new Event. I think this could be added without major implications to the consensus layer, and is in parallel to the Tendermint "filter in mempool" approach. However, one a bad Tx is witnessed by a few other members and buried in a tree of Tx, if a new member refuses to accept the entire tree based on the Tx, this could break the network. And thus this CheckTx would only be useful when there was exactly one malicious member. Two of them could bury an invalid Tx below various valid Tx and then force it into a block anyway

Spot on. I am reticent to add the CheckTx on direct parents because it doesn't actually prevent a determined spammer. So it feels like a massive overhead that doesn't fix the intended problem. As you mention, doing it properly would require waiting for entire trees with 2/3 of participants before finalizing an Event. This in turn requires a mechanism to rollback "bad" branches, which is obviously very tricky. 

So yes, pending further reflection. I'll close the issue for now.
